Mu2eG4/inc/Mu2eG4PrimaryType.hh

```diff
@@ -0,0 +1,56 @@
+// Mu2eG4 can create G4 primaries from different Mu2e data types.
+// We want to use enums to represent those types in the code, but also
+// need a string representation to read them from fhicl config.
+//
+// An instance of the Mu2eG4PrimaryType class can be created from
+// either the enum or a string, and has both enum and string-typed
+// accessors.
+//
+// Implementation features:
+//     - thread safe
+//     - does not use heap
+//     - a single list of entities, no other lists to keep in sync
+//
+// Andrei Gaponenko, 2021
+
+
+#ifndef Mu2eG4_inc_Mu2eG4PrimaryInput_hh
+#define Mu2eG4_inc_Mu2eG4PrimaryInput_hh
+
+#include <string>
+#include <vector>
+
+
+// Master list of all Mu2eG4 primary input types
+#define MU2EG4_PRIMARY_TYPES                         \
+  X( GenParticles )                                  \
+  X( StepPoints )                                  \
+  /**/
+
+namespace mu2e {
+
+  class Mu2eG4PrimaryType {
+  public:
+   enum enum_type {
+#define X(x) x,
```

Comment:
Can you comment why you are using macro programming instead of (say) defining a templated function here?  What problem are you trying to solve?

Reply:
Hi Dave,

> I'd like to understand why preprocessor macros are being used; my
> experience is these are hard to debug and cryptic ('X' is not a
> very expressive name).

Yes, C preprocessor macros can be easily abused.
(A recent fiasco with "#define time ...."  in a G4 header file comes to mind.)

However they are a non-deprecated feature of the current C++ language,
because some of the provided functionality has no equivalent
replacement.  In this PR the macro in question is used to interpret
the same sequence of characters in the source file as a literal or a
string.  This allows us to maintain a single list of input types: just
the MU2EG4_PRIMARY_TYPES macro will need to be updated.  Contrast this
with the alternative approach of maintaining two lists in sync and
dealing with patches like the one to MCDataProducts/inc/ProcessCode.hh
here

https://github.com/Mu2e/Offline/pull/308/commits/a3c99ec9c87d64cf1a4ea05d94d695555f60f30d#diff-d2901991f9847a8d79b64710fd2d3864e30aff03e25cb9d5e53519e80cdb607f

It is not easy to check whether the update would get the lists out of
sync.  (And the list of strings there is a preprocessor macro anyway!)

The macro name 'X' is conventional in the "X macro technique" that
is used here, see https://en.wikipedia.org/wiki/X_Macro

Reply:
Thanks for the reference.  I understand the potential advantage of what this does, though for a list of 2 values it seems heavy.  To help future coders recognize this pattern please include the link above and the comments about how this works and should be maintained in the code.

Reply:
A comment has been added in commit 694e28cacc8b1c6b89939e275795b5532a236981